feat(radiation): radiation rework :goberserk:

igorsaux opened the pull request:
Был выбран вариант `Попытаться сделать полноценную систему с альфа-, бета- и гамма-излучениями, где, как и ирл, каждый тип излучения по-своему уникален.`

close #8551
close #1119

<details>
<summary>Чейнджлог</summary>

```yml
🆑
rscadd: Добавлена приближенная к реальности симуляция радиации.
tweak: Счётчик Гейгера отображает поглощаемую дозу в греях и отображает все виды излучении в его зоне.
bugfix: Коллекторы радиации теперь работают только на излучении Хокинга.
rscadd: Сингулярность теперь излучает частицы Хокинга.
balance: Коллекторы радиации постепенно теряют здоровье при получении дозы выше 1 Грей.
/🆑
```

</details>

- [x] Pull Request полностью завершен, мне не нужна помощь чтобы его закончить.
- [x] Я внимательно прочитал все свои изменения и багов в них не нашел.
- [x] Я запускал сервер со своими изменениями локально и все протестировал.
- [x] Я ознакомился c [Guide to Contribute](https://github.com/ChaoticOnyx/OnyxBay/blob/dev/docs/contributing.md).

Voldirs commented:
нииихуя себе

kemperhun commented:
Добавил космонавтам массу, но не прикрутил ее к киллерфиче билда, эх...

O0oMaximo0O commented on code/controllers/subsystems/radiation.dm:
```diff
@@ -49,53 +35,58 @@ SUBSYSTEM_DEF(radiation)
-		var/dist = get_dist(source.source_turf, T)
-		if(dist > source.range)
-			continue // Too far to possibly affect
 		if(source.flat)
```
Точно для плоской расстояние мерять не надо?

O0oMaximo0O commented on code/game/objects/items/devices/geiger.dm:
```diff
@@ -12,31 +8,72 @@
 		return
-	radiation_count = SSradiation.get_rads_at_turf(get_turf(src))
+
+	radiation_dose = 0
+	radiation_activity = 0
```
Что-то фигня какая-то.
По идее счётчик Гейгера должен работать в двух режимах, судя по выводу сообщения ниже
```
var/msg = "[scanning ? "Ambient" : "Stored"] Radiation: [fmt_siunit(radiation_dose, "Gy", 3)].<br>"
```
В зависимости от режима он должен либо накапливать дозу радиации, либо фиксировать текущий фон.
А тут это игнорится. В итоге есть две обнуляемые характеристики, а счётчик просто либо работает, либо нет.

O0oMaximo0O replied:
Ладно, это отдельный иссуй и ПР нужен.

O0oMaximo0O commented on code/modules/mob/living/carbon/brain/life.dm:
```diff
@@ -2,35 +2,30 @@
+		return
+
+	if(radiation > (3 SIEVERT))
+		radiation = 3 SIEVERT
```
Зачем ограничивать 200, если ниже до 3 сбрасывает?

O0oMaximo0O commented on code/modules/mob/living/carbon/human/life.dm:
```diff
@@ -212,70 +212,63 @@
 		if(species.appearance_flags & RADIATION_GLOWS)
-			set_light(0.3, 0.1, max(1,min(20,radiation/20)), 2, species.get_flesh_colour(src))
-		// END DOGSHIT SNOWFLAKE
+			set_light(0.3, 0.1, max(1,min(20, radiation * 25)), 2, species.get_flesh_colour(src))
```
```suggestion
			set_light(0.3, 0.1, Clamp(radiation * 25, 1, 20), 2, species.get_flesh_colour(src))
```

O0oMaximo0O commented on code/game/turfs/turf.dm:
```diff
@@ -26,6 +26,12 @@
 	var/icon_old = null
 	var/pathweight = 1          // How much does it cost to pathfind over this turf?
 	var/blessed = 0             // Has the turf been blessed?
+	var/list/rad_resist = list(
+		RADIATION_ALPHA_RAY = 0.9,
+		RADIATION_BETA_RAY = 0.5,
+		RADIATION_GAMMA_RAY = 0.1,
+		RADIATION_HAWKING_RAY = 0.4
+	)
```
Так, у нас что, пол тоже сдерживать радиацию будет? О_О

igorsaux replied:
не пол а атмосфера

O0oMaximo0O replied:
Не слишком круто? Хотя бы в сотые переделай.

Voldirs replied:
Вообще альфа останавливается даже кожей и для человека альфа-излучение опасно только в том случае если его источник попадёт внутрь человека, в желудок, в кровь или в лёгкие в виде инертного газа радона. Но в этом случае альфа излучение с его энергией оказывает ооооооочень серьёзное влияние на организм, тут оно опаснее и гаммы, и беты будет

O0oMaximo0O commented on code/modules/radiation/radiation.dm:
```diff
@@ -1,12 +1,103 @@
+
+/// Returns J
+/datum/radiation_info/proc/calc_energy_rt(atom/source, atom/target)
+	var/result_energy = energy
+	var/turf/target_turf = get_turf(target)
+
+	while(source.loc != target_turf.loc)
+		var/dist = get_dist(source, target_turf)
+
+		ASSERT(dist > 0)
+
+		var/turf/source_turf = get_turf(source)
+
+		result_energy = result_energy / (dist ** 2)
+		result_energy -= (result_energy * source_turf.calc_rad_resistance(src))
+
+		if(result_energy < INSUFFICIENT_RADIATON_ENERGY)
+			return 0
+
+		source = get_step_towards(source, target_turf)
+	
+	return result_energy
```
```suggestion
	var/turf/current_turf = get_turf(source)
	ASSERT(current_turf)
	var/turf/target_turf = get_turf(target)
	ASSERT(target_turf)

	var/unabsorbed_energy_k = 1.0
	while(current_turf != target_turf)
		unabsorbed_energy_k *= 1.0 - current_turf.calc_rad_resistance(src)
		if(unabsorbed_energy_k <= 0)
			return 0
		current_turf = get_step_towards(current_turf, target_turf)
	var/result_energy = energy * unabsorbed_energy_k / get_dist(source, target_turf) ** 2
	if(result_energy < INSUFFICIENT_RADIATON_ENERGY)
		return 0
	return result_energy
```

O0oMaximo0O commented on code/modules/radiation/radiation.dm:
```diff
@@ -1,59 +1,315 @@
+	ASSERT(source.loc)
+	ASSERT(target.loc)
+
+	if(isobj(current_point))
+		var/obj/current_obj = current_point
+		result_energy -= (result_energy * current_obj.calc_rad_resistance(src))
+		current_point = current_point.loc
+	
+	result_energy = result_energy / (get_dist(get_turf(source), get_turf(target)) ** 2)
+
+	if(result_energy < INSUFFICIENT_RADIATON_ENERGY)
+		return 0
+
+
+	// Example of traverse: [beaker] -> [box] -> [human] -> [turf]
+	while(!isturf(current_point) && current_point != target)
+		if(!isobj(current_point) || current_point == source)
+			current_point = current_point.loc
+			continue
+
+		var/obj/current_obj = current_point
+		result_energy -= (result_energy * current_obj.calc_rad_resistance(src))
+
+		if(result_energy < INSUFFICIENT_RADIATON_ENERGY)
+			return 0
+		
+		current_point = current_point.loc
```
Дублирование кода.

O0oMaximo0O commented on code/modules/radiation/radiation.dm:
```diff
@@ -1,59 +1,315 @@
+				target_parent = target_parent.loc
+
+			return result_energy
+		// Not on the same turf
+		else
+			var/turf/current_turf = get_turf(current_point)
+			var/on_same_turf = get_turf(source) == current_turf
```
Странная проверка после такого комментария)

O0oMaximo0O replied:
Лучше поместить цикл раскрутки таргета наружу, до или после прохода по тайлам.

TermService commented on code/__defines/flags.dm:
```diff
@@ -19,6 +19,7 @@ GLOBAL_LIST_INIT(bitflags, list(1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 204
 #define ATOM_FLAG_INITIALIZED            0x0020 // Has this atom been initialized
 #define ATOM_FLAG_FULLTILE_OBJECT        0x0040 // Blocks interactions with most atoms on the same tile, except mobs, items and anything that has the flag below, i.e. fulltile windows
 #define ATOM_FLAG_ADJACENT_EXCEPTION     0x0080 // Skips adjacent checks for atoms that should always be reachable in window tiles
+#define ATOM_FLAG_IGNORE_RADIATION       0x0100 // It will not produce any radiation when it will be a radiation source.
```
160 же не?

igorsaux replied:
не

Tsurupeta replied:
Игорь, может в битшифты переделаешь?

MrMagick2104 commented on code/modules/power/singularity/collector.dm:
```diff
@@ -36,17 +35,33 @@ var/global/list/rad_collectors = list()
+
+			var/exceed = max(0.0, total_dose - (1 GREY))
+			if(exceed > 0)
+				health -= (exceed * 2)
```
Хардкап выработки коллекторов, который ещё их и взрывает - очень плохо.
Уже проходили в https://github.com/ChaoticOnyx/OnyxBay/pull/3854.
Текущие константы не подходят для сингулярности, ибо она взрывает коллекторы, это понятно, но, может, это вообще не нужно?

Commit:
feat(radiation): radiation rework


turfs and obj resist


mob balance


code tweaks


retard bugs


fix flat radiation and damage


fix ebli


gg

TermService reviewed (dismissed):
Ладно

Commit:
hawking resist

Doster-d reviewed (approved):
Тесты успешно, игроки недовольны (нет), можно в прод
